### ISSUE TITLE
Idea for generics array w/o performance impact

### ISSUE THREAD

**Issue by zmitic** [mindplay-dk/php-generics-rfc#46]
In lots of places I have seen generic array being the biggest performance killer; if parameter is typed as ``array<string>``, engine would need to check each member. And if same array is sent to another function, each member is checked again.

Is it possible to use different approach:
```php
function process(array<string> $p) {    // <- fatal error here
}
$data = array('php', 'rules');    // array of strings but without explicit <string> generic
process($data);
```

and get this error:

>_Fatal error: Uncaught TypeError: Argument 1 passed to process() must be of the type **array\<string\>**, **array\<mixed\>** given_

That is my suggestion; untyped array variable would be consider __internally__ as ``array<mixed>`` and thus, would not pass typehint in ``process()`` function; no need to manually check members.

##### Update: Error message doesn't have to say ``array<mixed>``, it can be ``array`` only. This is for illustration purposes only. 

To fix the above error, user only needs this:

```php
function process(array<string> $p) {
}

$data = array<string>('php', 'rules');    // declaration of type fixes the problem
process($data);
```

This solution would not loose runtime type checks (the reason for this issue); example:

```php
function process(array<string> $p) {
}

$data = array<string>('php', 'rules');  
$data[] = 42;    // <- fatal error here
process($data);
```

>_Fatal error (new for PHP8): int assigned to  **array\<string\>** in index.php line 5_

If the above is not possible in PHP, please close the issue and ignore the rest.

---
Above examples doesn't go into co(ntra)variance and coercive/strict mode (it would work as it works now, consider above example in strict mode). 
 
The big job is fixing methods like this (simplified examples):

```php
array_map ( callable $callback , array $array1 ) : array
// to
array_map<T=mixed> ( callable $callback , array<T> $array1 ) : array<T>

array_filter ( array $array callable $callback) : array
// to
array_filter<T=mixed> ( array<T> $array, callable $callback) : array<T>
```

I don't think this brings any BC problems. Application code would either use or not use generics, it won't matter; mixing both doesn't make sense. I am also sure that psalter/rector would provide automated upgrade.

Third party libs could (?) safely upgrade to returning generic arrays without breaking existing code. Even if it does break it, limiting version in composer.json would easily fix it.

**Comment by mindplay-dk** [comment]
> _Fatal error: Uncaught TypeError: Argument 1 passed to process() must be of the type **array\<string\>**, **array\<mixed\>** given_

Personally, I'm all for more strictness wherever possible, but this seems overly strict and not very "php" - it's a (very) dynamic language after all, and I'm not sure it's okay to let optional, gradual typing get in the way of the (lazy, sloppy) working man PHP programmer's productivity.

PHP is already partially gradually-typed, and already has aspects of similar performance problems - for example, if you were to implement a custom collection-type with a type-hinted append-method, then yes, everything you add to that collection gets type-checked whether or not that's actually necessary.

I think that's just the price difference you pay for using a gradually-typed language vs a compiled, statically type-checked language, where everything is deemed safe or not up ahead of run-time.

> I don't think this brings any BC problems. Application code would either use or not use generics, it won't matter; mixing both doesn't make sense.

The concern is not so much for application code, but for library code - no doubt, some libraries will incorporate generics, and some libraries won't or can't. This shouldn't lead to a divide in the ecosystem, where some libraries can be used in projects with generics and others can't.

**Comment by jkufner** [comment]
An another example to think about:

`````php
class Foo {
    public function __toString(): string { return "foo"; }
}
function bar(array<string> $a) { var_dump($a); }
$a = array<string>(new Foo(), "bar");
var_dump($a);
bar($a);
`````

**Comment by zmitic** [comment]
@mindplay-dk 
One-time check for ``Collection::add`` or ``$a[]='foo'`` isn't the problem, I created this issue only as idea for avoiding checking members of array **each** time they are used as param. Nothing more than that.

@jkufner This would fail as ``new Foo()`` is object, not a string. It _might_ work in coercive mode and $a would become

```array<string>('foo', 'bar')```

but that can be discussed later, when we see if this idea is even possible in PHP.

**Comment by mindplay-dk** [comment]
> I created this issue only as idea for avoiding checking members of array each time they are used as param. Nothing more than that.

But type-checking array elements would only happen when passing an untyped `array` to e.g. `array<string>` in any case, right?

There's no reason to type-check the elements if e.g. an `array<string>` is passed where an `array<string>` is expected, right?

Hmm, I wonder how this is going to jive with references... If you pass an untyped `array` to a by-reference `array<string>` parameter, what happens? There's probably nothing meaningful you can do here. So for that reason, what you're suggesting might actually be necessary - or if not, coercion would be somewhat inconsistent in this manner... 🤔

**Comment by zmitic** [comment]
>But type-checking array elements would only happen when passing an untyped array to e.g. array<string> in any case, right?

No, checking each member is **never** executed in function signature because ``array`` is not the same type as ``array<string>``. That is why I used same Fatal Error message we already have; it is **exactly** like passing string to a function expecting object or passing instance of ``User`` to a function expecting ``Category``; you immediately get error.

So engine only needs to check internal type, just like it does now. 3rd example shows how we still keep runtime checks, but only once per-member, not each time array is used (which is the problem everyone mentions on talks when asked about generics).

> If you pass an untyped array to a by-reference array<string> parameter, what happens

Fatal Error because just like above; ``array`` is not the same type as ``array<string>``. 

But to be honest, I can't remember last time I used pass-by-reference of arrays :smile: so I might be wrong.

>This shouldn't lead to a divide in the ecosystem, where some libraries can be used in projects with generics and others can't.

It won't; if application code uses library with generics, that application code will typehint generic arrays. If library doesn't use generics, application code relying on it won't use generics as well.

My first comment explained potential problem with this; all user has to do is to lock versions in composer.json and/or use psalter/rector to upgrade the code. If mixed libraries are used, ``@param array<string> $p`` would be used in application code.

#### Note 
co(ntra)variance of parameter/return type is not part of this issue.

**Comment by zmitic** [comment]
Another example why I think this won't bring BC problems when existing code without generics adds a library that uses them.

Imagine that some library is updated and uses this signature:

```php
// third-party lib
function getRandomNames(): array<string> // one-time check of return type (not members)
{
    return array<string>('John', 'Smith'); // one-time check per member
}
```


If application code **was** this

```php
function process(array $p) { // not typed array
}
$data = getRandomNames(); // engine knows this is array<string>, not a simple array
process($data);
```

code would keep working because ``array $p`` in ``process`` is not typed array thus **no** need to do any checks.

In other words; if parameter is typed array, only **then** do a check I described in first comment. This will keep BC.

So a user decides to update application code:

```php
function process(array<string> $p) { // one-time check of type, not of members
}
$data = getRandomNames(); // engine knows this is array<string>
process($data);
```

everything would work with minimal performance penalty because members **have been**  checked in ``getRandomNames()``.

**Comment by someniatko** [comment]
It looks like an introduction of a typed variable concept. IMO if you decide to allow specification of variable's type, you should go ahead and allow any types, not only array<T>.

**Comment by zmitic** [comment]
@someniatko 
>It looks like an introduction of a typed variable concept.

It isn't, my suggestion falls into generics declaration. Just like you would use generics for objects like 

```$a = new ArrayCollection<string>('a', 'b')```

you would do the same for array 

```$a = array<string>('a', 'b')```.

**Comment by mindplay-dk** [comment]
> Another example why I think this won't bring BC problems when existing code without generics adds a library that uses them.

The opposite case is what concerns me.

If the *library* code is updated from this:

```php
/**
 * @param array<string>
 */
function process(array $p) {
}
```

To this:

```php
function process(array<string> $p) {
}
```

And your *application* code was this:

```php
function getRandomNames(): array
{
    return array('John', 'Smith');
}

process(getRandomNames());
```

Now your application code is broken, right?

So just migrating from doc-blocks to proper type-hints in existing libraries will be a breaking change, and that's not really natural or expected, since changing an argument from "list of strings" to "type-checked list of strings" doesn't change the function or requirements at all, it just adds an extra layer of safety that you might even previously have been implementing by hand.

This situation is different from the addition of property type-hints, because properties don't appear in interfaces, or in function signatures.

It's a very different situation from adding something that affects the public surface area of any type.

**Comment by zmitic** [comment]
> Now your application code is broken, right?

You are right, it will get broken. 

But as I described it, user needs to either lock composer.json or upgrade the code. Because it is new language feature, for new **major** version of PHP (8), it is allowed and in this case, very easy to avoid.

Also if library gets this drastic change, author will surely put it as major version; composer would not even download it without user explicitly saying so.

> will be a breaking change, and that's not really natural or expected

Disagreed: [this implemented rfc](https://wiki.php.net/rfc/consistent_type_errors) will make serious BC problems in **all** internal functions, yet everyone is in favor and no arguing here.

---
So to conclude: BC discussion of something that doesn't even exist, for new major PHP version, are kinda moot. I am primarily interested if this is even **technically** possible to do in PHP; we can discuss other things later because if it is not possible, no need for any other discussion .

**Comment by mindplay-dk** [comment]
> Because it is new language feature, for new major version of PHP (8), it is allowed and in this case, very easy to avoid.

You shouldn't need to lock your dependencies with `composer.json` - that's not what that feature is for. Locking your dependencies in a project might work, but locking your dependencies in a library only helps that library and nobody who uses that library.

So any library that adds type-hints would have to version that as a breaking change.

Upgrading the language version requirement of a library doesn't automatically mean you have to version that change to the library as a breaking change - you shouldn't need to version a library as breaking unless it's API has changed. 

It's hard to argue the API has changed just because you added more type-checks, when the API is otherwise identical and accepts/returns the exact same data-structures.

> this implemented rfc will make serious BC problems in all internal functions

That's permitted in a major version of the language - those don't happen real often... so yeah, if generics was by some miracle to take form before PHP 8 is released, you could argue this breaking change is also in order.

However, as I said before, I'm not sure this is really in the spirit of PHP - being gradually-typed, I'm not sure it's really natural or expected to have strict type-checks for arrays and fail on incompatible calls only because the declared types don't match, even when the array contents *would* match.

That sort of thing is very likely to get a thumbs down from internals.

**Comment by korenevskiy** [comment]
Dear participants.
All the problems have already been discussed. It became clear how that should be.
I will take the liberty of describing the characteristics.
1. The `Array` type is not an `Array<string>`  type.
2. The `Array` type similar to the type `Array<mixed>`  type.
3. Checking of type elements will occur only when assigning an element by index, and when creating an array.
4. The emergence of generics introduces the need for new features. Since the elements are checked during assignment, there is a need for functions:
```
function array_settype<T>(array $a): array<T>;
//or 
function array_type(array $a,'T'): array<T>;
//and
function array_checktype<T>(array $a): bool;
//or 
function is_array ( mixed $var, 'T') : bool;
```
5. Modification of the `array_map` function there is an error,
```
array_map ( callable $callback , array $array1 ) : array
// to
array_map<T=mixed> ( callable $callback , array<T> $array1 ) : array<T>
```
Here the check can occur in specifying the type of parameter of the called function.
And since this function can return an array with a completely different type.
`array_map(function(string $el){return (int)strlen($el);},$a);`
We must specify the type for the function only for the return type.
```
array_map<T=mixed>( callable $callback , array $array1 ) : array<T>.
//or
array_map(callable $callback , array $array1 , 'T') : array<T>.
//With this call, the callback function can get the last parameter of the type string.
//This will allow you to specify the return type of the callback in the function itself.
//Already in the current next version of PHP, it is desirable to add this function.
```

Professional application developers use type checks every time in their code. It is obvious that checking the types of arrays at the PHP level will be performed faster than checking in the code of programmers.
The code will be cleaner.

6. With the introduction to genetics, there is a need to use conversions.
```
$a1 = ['foo','bar']; // type array<mixed>
$a2 = (array<string>)$a1; / / convert to type array<string>
```
which is equivalent to
```
function array_settype<T>(array $a): array<T>;
//or 
function array_type(array $a,'T'): array<T>;
```
and should call type conversion for each element.

7. Casting the `array<T>` type to the `array` type, as an exception, will pass without iterating through the array
```
$a1 = array<string>('foo','bar'); // type array<string>
$a2 = (array)$a1; / / convert to type array<mixed>
``` 
Should type conversion without call for each element.

8. It is worth introducing a new syntax with square brackets.
```
$a = string['foo','bar'];    //justified
$a = int[23,44];           //justified
$a = MyClass[];                //unnecessary
$a = MyClass[$myclass1, $myclass2]; //unnecessary
```
i think that using the new syntax to create objects is unnecessary for the introduction of generics. But creating to create arrays for simple types is quite justified.

9.For objects, you should enter the magic method **`__toType('T')`**.
which will return an object of the desired type for casting types when iterating through or when checking the types of the entire array. This method can return an object of the parent type of the inherited class.

10.Personally, i want to add an amendment.
`<?= //"Comment Text" ?>`
so that it doesn't cause an error.

11.When checking types, the inherited types must be checked. whether they match the type and whether they match the parent type specified in the function parameters.
```
class A{}
class B extends A{}
function MyFunc(array<A> $myclass){
}
$a = array<B>();
MyFunc($a);
```

12.Already in the current next version of PHP, you need add one function
```
function is_array ( mixed $var, 'T') : bool;
```
which, in principle, will solve some of the problems of checking arrays in the code of current developers today. This will give you some backward compatibility for common PHP libraries for future versions.
```
function is_array ( mixed $var) : bool;
```
Should treat arrays with any default types as arrays.

**PS**
Does it make sense in a future version of php using generics to use 2 kinds of syntax for one function?
```
function is_array ( mixed $var, 'T') : bool;
//and
function is_array<T> ( mixed $var) : bool;
```
In the first case, it will be backward compatible with the current next version of PHP, which is convenient for general-purpose public libraries. Provided that this method appears already in the current version of PHP.
in the second case, it is 100% typed code, with ide typing support.
